lib/ui/window.dart

```diff
@@ -286,31 +569,212 @@ class Locale {
+          if (entry.value.isNotEmpty)
+            resultVWYZX.write('-${entry.value}');
+        } else if (entry.key != 'x') {
+          throw UnimplementedError('Unexpected key in extensions map: $entry');
+        }
+      } else if (_reKey.hasMatch(entry.key)) {
+        // unicode_locale_extensions
+        if (entry.value == 'true' || entry.value == '') {
+          uOut.write('-${entry.key}');
+        } else {
+          uOut.write('-${entry.key}-${entry.value}');
+        }
+      } else if (_reTkey.hasMatch(entry.key)) {
+        // transformed_extensions
+        tOut.write('-${entry.key}');
+        // TODO: this is not standards compliant. What do we want to do with
+        // this case? Drop entry.key like we drop empty t and u singletons?
+        // Or simply ensure we don't ever create such an instance?
+        if (entry.value.isNotEmpty)
+          tOut.write('-${entry.value}');
+      } else {
+        throw UnimplementedError('Unexpected key in extensions map: $entry');
+      }
+    }
+    if (tLang != null || tOut.isNotEmpty) {
+      result.write('-t');
+      if (tLang != null)
+        result.write('-$tLang');
+      result.write(tOut.toString());
+    }
+    if (uAttributes != null || uOut.isNotEmpty) {
+      result.write('-u');
+      if (uAttributes != null && uAttributes.isNotEmpty)
+        result.write('-$uAttributes');
+      result.write(uOut.toString());
+    }
+    if (resultVWYZX.isNotEmpty)
+      result.write(resultVWYZX.toString());
+    if (extensions.containsKey('x')) {
+      result.write('-x');
+      if (extensions['x'].isNotEmpty) {
+        result.write('-${extensions["x"]}');
+      }
+    }
+    return result.toString();
+  }
+
+  // Unicode Language Identifier subtags
+  // TODO/WIP: because we lowercase Locale Identifiers before parsing, typical
+  // use of these regexps don't actually need to atch capitals too.
+  static final RegExp _reSingleton = RegExp(r'^[a-zA-Z]$');
+
+  // (https://www.unicode.org/reports/tr35/#Unicode_language_identifier).
+  static final RegExp _reLanguage = RegExp(r'^[a-zA-Z]{2,3}$|^[a-zA-Z]{5,8}$');
+  static final RegExp _reScript = RegExp(r'^[a-zA-Z]{4}$');
+  static final RegExp _reRegion = RegExp(r'^[a-zA-Z]{2}$|^[0-9]{3}$');
+  static final RegExp _reVariant = RegExp(r'^[a-zA-Z0-9]{5,8}$|^[0-9][a-zA-Z0-9]{3}$');
+  static final RegExp _reSep = RegExp(r'[-_]');
+
+  // Covers all subtags possible in Unicode Locale Identifiers, used for
+  // pu_extensions.
+  static final RegExp _reAllSubtags = RegExp(r'^[a-zA-Z0-9]{1,8}$');
+  // Covers all subtags within a particular extension, used for other_extensions.
+  static final RegExp _reOtherSubtags = RegExp(r'^[a-zA-Z0-9]{2,8}$');
+  // Covers "attribute" and "type" from unicode_locale_extensions, and "tvalue" in
+  // transformed_extensions.
+  // (https://www.unicode.org/reports/tr35/#Unicode_locale_identifier).
+  static final RegExp _reValueSubtags = RegExp(r'^[a-zA-Z0-9]{3,8}$');
```

Comment:
Assume I've lowercased the string, would a for loop with manual codeUnitAt checks for a-z and 0-9 still be faster? (I shouldn't bring my distrust of scripting language for-loop speed into Dart, because Dart is typically compiled, or for Web, JavaScript for loops are fast enough these days?)

Reply:
I would test it. I do expect the for loop to be faster.
You can check: https://dartpad.dartlang.org/157081da0debce48c43d771da637b7fb
which compares `RegExp(r"^[a-zA-Z\d]{1,8}$")` to a hand written function.

The loop is 80% faster in the browser, ~200% faster on the VM, when benchmarked almost entirely on accepted strings. That is, it's the time taken for valid input.